### PR TITLE
Cms duplicate

### PR DESCRIPTION
**Pull request by Pikkupomo**
Ensure `Page` outputs text from `__str__`
CMS: No duplicate URLs allowed from now on
Refs SHOOP-1571

### CODE REVIEW

**Comment by akx** [comment]
Can you rewrite the latter commit message to match the usual style? 

Also prefix the first commit message with "CMS:" too :)

File: shoop/simple_cms/admin_module/views.py [hunk 1]
```diff
@@ -69,6 +73,14 @@ def clean(self):
 
         return data
 
+    def validate_url(self, field_name, url):
```
Comment (akx) [inline shoop/simple_cms/admin_module/views.py]: Docstring, please.

File: shoop/simple_cms/admin_module/views.py [hunk 2]
```diff
@@ -69,6 +73,14 @@ def clean(self):
...
+        qs = self._get_translation_model().objects.filter(url=url)
+        if self.instance.pk:
+            qs = qs.exclude(master=self.instance)
+
+        if qs.exists():
+            self.add_error(field_name, _("URL already exists."))
```
Comment (akx) [inline shoop/simple_cms/admin_module/views.py]: Use error codes -- they're easier to test.

```python
self.add_error(field_name, ValidationError(_(bla), code="duplicate_url")
```
...
```
assert form.errors["url__en"][0].code == "duplicate_url"
```

File: shoop_tests/simple_cms/test_page_creation.py [hunk 3]
```diff
@@ -59,9 +59,31 @@ def test_page_form(rf):
...
     # Let's edit that page
-    data.update({"title__en": "englaish", "url__en": "errrnglish", "content__en": "ennnn ennnn ennnnnnn-nn-n-n"})
+    original_url = "errrnglish"
+    data.update({"title__en": "englaish", "url__en": original_url, "content__en": "ennnn ennnn ennnnnnn-nn-n-n"})
     form = form_class(**dict(form_kwargs, data=data, instance=page))
     form.full_clean()
     assert not form.errors
     page = form.save()
     assert set(page.get_available_languages()) == {"fi", "en"}  # English GET
+
```
Comment (akx) [inline shoop_tests/simple_cms/test_page_creation.py]: Could this be another test case?

File: shoop/simple_cms/admin_module/views.py (hunk 2 above)
Comment (Pikkupomo) [inline shoop/simple_cms/admin_module/views.py]: `BaseForm.add_error()` is defined as `def add_error(self, field, error)`

File: shoop/simple_cms/admin_module/views.py [hunk 4]
```diff
@@ -69,6 +73,36 @@ def clean(self):
...
+
+        for existing page:
+        * URL (other than owned by existing page) exists
+        * URL exists in other languages of existing page
+        """
+        qs = self._get_translation_model().objects.filter(url=url)
+
+        if not self.instance.pk:
+            if qs.exists():
+                self.add_error(field_name, _("URL already exists."))
```
Comment (akx) [inline shoop/simple_cms/admin_module/views.py]: I don't really see how it's this function's responsibility to add the error too (also, this same line is copypasted thrice).

Comment (akx) [inline shoop/simple_cms/admin_module/views.py]: Also, please use ValidationErrors with error codes.

File: shoop/simple_cms/admin_module/views.py (hunk 2 above)
Comment (akx) [inline shoop/simple_cms/admin_module/views.py]: `self.add_error(field_name, ValidationError(_(bla), code="duplicate_url"))`

You pass a `ValidationError` as the error.

**Comment by akx** [comment]
Yay, thanks!

**Pull request merged**